### PR TITLE
B2 gateway S3 compat: return MD5 hash as ETag from PutObject

### PR DESCRIPTION
**Pull request by KiNgMaR**
## Description

Quoting the B2 Gateway Readme:

> PutObject S3 API does not return the md5sum of an uploaded file as it's etag response. Applications that check the validity of the uploaded object by comparing these two values will fail.

However, as confirmed by the docs, and as tested in both US and EU regions, B2 actually does return an MD5 hash for a stored file via the https://www.backblaze.com/b2/docs/b2_get_file_info.html endpoint.

This we can use to provide better compatibility with S3 client libraries that assume the ETag is the MD5 hash - such as boto.

The change depends on a PR for blazer: https://github.com/kurin/blazer/pull/70

I opted to only enable this change if MinIO's `--compat` mode is active, however this can be discussed.

## Motivation and Context

I was unable to use MinIO for migrating about 700.000 files from local storage to B2 because of the error described below. This is a Seafile server, which uses the [boto library](https://github.com/boto/boto) in version 2.49, however the MD5-ETag requirement also seems to exist in boto3: https://github.com/boto/boto3/issues/544

**This PR is not ready to merge - I would like to get some feedback on whether this approach seems acceptable to you. Definitely missing are: unit tests, updated go.mod for blazer library.**

## How to test this PR?

- Run MinIO in B2 Gateway mode. Try to upload a file via a tool that uses boto. Run into the following error:
`BotoClientError: ETag from S3 did not match computed MD5. "4_...-1" vs. b29c551c2decea1c0625b34fda00db09
`
- This usually causes the file to exist in the B2 bucket because most clients don't remove it after running into this error.
- Apply patches and start MinIO with `--compat`.
- Re-try upload, should succeed without an error.

## Types of changes
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] Discussion of approach with `--compat` flag
- [ ] Unit tests needed
- [x] Updated go.mod with changes for blazer library

### CODE REVIEW

**Comment by minio-trusted** [comment]
```
Error running mint automation
+ mkdir -p 9183-369b5cb/gopath/src/github.com/minio
+ git -C 9183-369b5cb/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 9183-369b5cb/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 9183-369b5cb/gopath/src/github.com/minio/minio fetch --quiet minio pull/9183/head:pr9183
+ git -C 9183-369b5cb/gopath/src/github.com/minio/minio checkout --quiet pr9183
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/9183-369b5cb/gopath
+ make -C 9183-369b5cb/gopath/src/github.com/minio/minio --quiet
go: downloading github.com/aliyun/aliyun-oss-go-sdk v0.0.0-20190307165228-86c17b95fcd5
go: downloading github.com/minio/cli v1.22.0
go: downloading google.golang.org/api v0.5.0
go: downloading github.com/minio/hdfs/v3 v3.0.1
go: downloading github.com/minio/minio-go/v6 v6.0.51-0.20200319192131-097caa7760c7
go: downloading cloud.google.com/go v0.39.0
go: downloading github.com/json-iterator/go v1.1.9
go: downloading github.com/minio/gokrb5/v7 v7.2.5
go: downloading github.com/tinylib/msgp v1.1.1
go: downloading github.com/lib/pq v1.1.1
go: downloading gopkg.in/olivere/elastic.v5 v5.0.80
go: downloading golang.org/x/crypto v0.0.0-20191117063200-497ca9f6d64f
go: downloading github.com/Shopify/sarama v1.24.1
go: downloading github.com/gorilla/handlers v1.4.0
go: downloading github.com/valyala/tcplisten v0.0.0-20161114210144-ceec8f93295a
go: downloading github.com/Azure/azure-pipeline-go v0.2.1
go: downloading golang.org/x/net v0.0.0-20190923162816-aa69164e4478
go: downloading github.com/kurin/blazer v0.5.4-0.20190613185654-cf2f27cc0be3
go: downloading github.com/klauspost/compress v1.10.3
go: downloading github.com/gorilla/rpc v1.2.0
go: downloading google.golang.org/genproto v0.0.0-20190513181449-d00d292a067c
go: downloading github.com/modern-go/reflect2 v1.0.1
go: downloading github.com/philhofer/fwd v1.0.0
go: downloading github.com/ncw/directio v1.0.5
go: downloading github.com/dustin/go-humanize v1.0.0
go: downloading github.com/prometheus/client_golang v0.9.3-0.20190127221311-3c4408c8b829
go: downloading github.com/elazarl/go-bindata-assetfs v1.0.0
go: downloading github.com/nsqio/go-nsq v1.0.7
go: downloading github.com/mailru/easyjson v0.0.0-20180730094502-03f2033d19d5
go: downloading github.com/klauspost/readahead v1.3.1
go: downloading github.com/minio/lsync v1.0.1
go: downloading gopkg.in/ini.v1 v1.48.0
go: downloading github.com/skyrings/skyring-common v0.0.0-20160929130248-d1c0bb1cbd5e
go: downloading github.com/minio/highwayhash v1.0.0
go: downloading github.com/coreos/etcd v3.3.12+incompatible
go: downloading github.com/golang/protobuf v1.3.2
go: downloading golang.org/x/sys v0.0.0-20190922100055-0a153f010e69
go: downloading github.com/go-sql-driver/mysql v1.4.1
go: downloading github.com/klauspost/reedsolomon v1.9.3
go: downloading github.com/coredns/coredns v1.4.0
go: downloading github.com/googleapis/gax-go/v2 v2.0.4
go: downloading github.com/prometheus/common v0.2.0
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading github.com/shirou/gopsutil v2.18.12+incompatible
go: downloading github.com/cheggaaa/pb v1.0.28
go: downloading github.com/mattn/go-ieproxy v0.0.0-20190805055040-f9202b1cfdeb
go: downloading github.com/mattn/go-isatty v0.0.7
go: downloading github.com/golang/snappy v0.0.1
go: downloading github.com/rcrowley/go-metrics v0.0.0-20190704165056-9c2d0518ed81
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading github.com/beevik/ntp v0.2.0
go: downloading github.com/djherbis/atime v1.0.0
go: downloading github.com/minio/sha256-simd v0.1.1
go: downloading golang.org/x/text v0.3.2
go: downloading gopkg.in/jcmturner/gokrb5.v7 v7.2.3
go: downloading github.com/minio/sio v0.2.0
go: downloading github.com/rs/cors v1.6.0
go: downloading github.com/gomodule/redigo v2.0.0+incompatible
go: downloading github.com/prometheus/procfs v0.0.0-20190117184657-bf6a532e95b1
go: downloading github.com/cespare/xxhash/v2 v2.1.1
go: downloading github.com/secure-io/sio-go v0.3.0
go: downloading github.com/dgrijalva/jwt-go v3.2.0+incompatible
go: downloading github.com/hashicorp/vault/api v1.0.4
go: downloading github.com/streadway/amqp v0.0.0-20190404075320-75d898a42a94
go: downloading golang.org/x/time v0.0.0-20190308202827-9d24e82272b4
go: downloading github.com/fatih/color v1.7.0
go: downloading google.golang.org/grpc v1.22.0
go: downloading github.com/beorn7/perks v0.0.0-20180321164747-3a771d992973
go: downloading github.com/Azure/azure-storage-blob-go v0.8.0
go: downloading github.com/hashicorp/vault/sdk v0.1.13
go: downloading go.opencensus.io v0.21.0
go: downloading github.com/hashicorp/go-uuid v1.0.1
go: downloading github.com/miekg/dns v1.1.8
go: downloading github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21
go: downloading github.com/hashicorp/go-retryablehttp v0.5.4
go: downloading github.com/klauspost/cpuid v1.2.2
go: downloading github.com/hashicorp/golang-lru v0.5.1
go: downloading gopkg.in/jcmturner/goidentity.v3 v3.0.0
go: downloading github.com/eclipse/paho.mqtt.golang v1.2.0
go: downloading github.com/mattn/go-runewidth v0.0.4
go: downloading github.com/eapache/queue v1.1.0
go: downloading github.com/jcmturner/gofork v0.0.0-20190328161633-dc7c13fece03
go: downloading github.com/nats-io/stan.go v0.4.5
go: downloading github.com/gogo/protobuf v1.2.1
go: downloading github.com/eapache/go-resiliency v1.2.0
go: downloading github.com/hashicorp/hcl v1.0.0
go: downloading github.com/minio/parquet-go v0.0.0-20200125064549-a1e49702e174
go: downloading github.com/nats-io/nuid v1.0.1
go: downloading github.com/hashicorp/go-multierror v1.0.0
go: downloading github.com/xdg/scram v0.0.0-20180814205039-7eeb5667e42c
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.1
go: downloading github.com/mattn/go-colorable v0.1.1
go: downloading github.com/hashicorp/errwrap v1.0.0
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading github.com/hashicorp/go-sockaddr v1.0.2
go: downloading golang.org/x/oauth2 v0.0.0-20190402181905-9f3314589c9a
go: downloading github.com/pkg/errors v0.8.1
go: downloading github.com/ryanuber/go-glob v1.0.0
go: downloading git.apache.org/thrift.git v0.13.0
go: downloading github.com/gorilla/mux v1.7.0
go: downloading gopkg.in/ldap.v3 v3.0.3
go: downloading github.com/rjeczalik/notify v0.9.2
go: downloading github.com/hashicorp/go-rootcerts v1.0.1
go: downloading go.uber.org/atomic v1.3.2
go: downloading github.com/tidwall/gjson v1.3.5
go: downloading github.com/pierrec/lz4 v2.4.0+incompatible
go: downloading github.com/klauspost/pgzip v1.2.1
go: downloading gopkg.in/square/go-jose.v2 v2.3.1
go: downloading github.com/inconshreveable/go-update v0.0.0-20160112193335-8152e7eb6ccf
go: downloading github.com/tidwall/match v1.0.1
go: downloading github.com/bcicen/jstream v0.0.0-20190220045926-16c1f8af81c2
go: downloading github.com/tidwall/pretty v1.0.0
go: downloading github.com/nats-io/nats.go v1.9.1
go: downloading github.com/xdg/stringprep v1.0.0
go: downloading gopkg.in/jcmturner/dnsutils.v1 v1.0.1
go: downloading github.com/nats-io/nkeys v0.1.3
go: downloading github.com/minio/simdjson-go v0.1.5-0.20200303142138-b17fe061ea37
go: downloading github.com/mitchellh/mapstructure v1.1.2
go: downloading github.com/alecthomas/participle v0.2.1
go: downloading github.com/prometheus/client_model v0.0.0-20190129233127-fd36f4220a90
go: downloading github.com/fatih/structs v1.1.0
go: downloading gopkg.in/jcmturner/rpc.v1 v1.1.0
go: downloading gopkg.in/yaml.v2 v2.2.4
go: downloading github.com/nats-io/jwt v0.3.2
go: downloading gopkg.in/jcmturner/aescts.v1 v1.0.1
go: downloading github.com/tidwall/sjson v1.0.4
go: downloading gopkg.in/asn1-ber.v1 v1.0.0-20181015200546-f715ec2f112d
go: downloading github.com/hashicorp/go-cleanhttp v0.5.1
# github.com/minio/minio/cmd/gateway/b2
cmd/gateway/b2/gateway-b2.go:259:27: fi.MD5 undefined (type *base.FileInfo has no field or method MD5)
cmd/gateway/b2/gateway-b2.go:260:12: fi.MD5 undefined (type *base.FileInfo has no field or method MD5)
make: *** [build] Error 2

```

File: docs/gateway/b2.md [hunk 1]
```diff
@@ -52,6 +51,7 @@ Gateway inherits the following B2 limitations:
 Other limitations:
 
 - Bucket notification APIs are not supported.
+- The PutObject S3 API only returns the md5sum of an uploaded file as it's etag response when minio is started in --compat mode. Use this only if you are stuck with a client library where this check cannot be turned off (e.g. boto).
```
Comment (harshavardhana) [inline docs/gateway/b2.md]: can you show on example where this cannot be turned off @KiNgMaR ? -

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by harshavardhana** [comment]
> I opted to only enable this change if MinIO's `--compat` mode is active, however this can be discussed.

Thanks this is a good idea @KiNgMaR

**Comment by harshavardhana** [comment]
Also please fix build issues locally @KiNgMaR using `make test, make install, make verify`

File: docs/gateway/b2.md (hunk 1 above)
Comment (KiNgMaR) [reply docs/gateway/b2.md]: Sure!
- for boto 2: https://github.com/boto/boto/blob/9e1cd3bd76e738d80630f1bd9160fd87c8eab865/boto/s3/key.py#L1000
- for boto 3: https://github.com/boto/boto3/issues/544

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#c37dabac-ce69-4e35-b083-acc09ab7e9b2) |
<a name="c37dabac-ce69-4e35-b083-acc09ab7e9b2"></a>
### 9183-d29c665/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:30587
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 851374d13cdc:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 8 minutes and 25 seconds
(4/15) Running aws-sdk-ruby tests ... done in 19 seconds
(5/15) Running awscli tests ... done in 2 minutes and 33 seconds
(6/15) Running healthcheck tests ... done in 1 seconds
(7/15) Running mc tests ... done in 6 minutes and 41 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 44 seconds
(9/15) Running minio-go tests ... done in 22 minutes and 1 seconds
(10/15) Running minio-java tests ... FAILED in 8 minutes and 16 seconds
{
  "name": "minio-java",
  "function": "composeObject(String bucketName, String objectName,List<ComposeSource> composeSources, Map <String,String > headerMap, ServerSideEncryption sseTarget)",
  "args": "size: 6 MB & 6 MB",
  "duration": 11252,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code=NotImplemented, message=A header you provided implies functionality that is not implemented, bucketName=minio-java-test-297j01j, objectName=minio-java-test-1c7abps, resource=/minio-java-test-297j01j/minio-java-test-1c7abps, requestId=16007C19CD5638C5, hostId=9c5a769d-b523-4ddb-a84c-312e728ba02d)\nrequest={method=PUT, url=http://minio-dev6.minio.io:30587/minio-java-test-297j01j/minio-java-test-1c7abps?uploadId=efee15fc622827ec&partNumber=1, headers=x-amz-copy-source-if-match: 56d0f49e86027344cad54e41b51acce6-2\nx-amz-copy-source: minio-java-test-297j01j/minio-java-test-28nm1sh\nAccept-Encoding: identity\nHost: minio-dev6.minio.io:30587\nUser-Agent: MinIO (amd64; amd64) minio-java/dev\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20200328T135159Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20200328/us-east-1/s3/aws4_request, SignedHeaders=accept-encoding;host;x-amz-content-sha256;x-amz-copy-source;x-amz-copy-source-if-match;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=501, headers=Accept-Ranges: bytes\nContent-Length: 410\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO/DEVELOPMENT.GOGET\nVary: Origin\nX-Amz-Request-Id: 16007C19CD5638C5\nX-Xss-Protection: 1; mode=block\nDate: Sat, 28 Mar 2020 13:51:59 GMT\n}\n >>> [io.minio.MinioClient.executeReq(MinioClient.java:1220), io.minio.MinioClient.execute(MinioClient.java:1082), io.minio.MinioClient.executePut(MinioClient.java:1446), io.minio.MinioClient.executePut(MinioClient.java:1468), io.minio.MinioClient.uploadPartCopy(MinioClient.java:2619), io.minio.MinioClient.composeObject(MinioClient.java:2572), FunctionalTest.composeObject_test1(FunctionalTest.java:2587), FunctionalTest.runTests(FunctionalTest.java:3396), FunctionalTest.main(FunctionalTest.java:3514)]"
}

Executed 9 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#8e7bebb0-4b8a-4e0c-8de5-f0e2e1804b99) |
<a name="8e7bebb0-4b8a-4e0c-8de5-f0e2e1804b99"></a>
### 9183-db6d71b/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:32414
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 54c9193a683a:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 8 minutes and 37 seconds
(4/15) Running aws-sdk-ruby tests ... done in 19 seconds
(5/15) Running awscli tests ... FAILED in 55 seconds
{
  "name": "awscli",
  "duration": 1533,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (BucketAlreadyOwnedByYou) when calling the CreateBucket operation: Your previous request to create the named bucket succeeded and you already own it."
}

Executed 4 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]

**Review by kannappanr — APPROVED** [review]
LGTM

**Pull request merged**